In [2]:
import glob
import os

import pandas as pd
import screed

from sourmash import signature as sig
from sourmash._minhash import hash_murmur


In [5]:
! aws s3 cp s3://kmer-hashing/tabula-muris/n_hashes=500/bladder/similarities_tidy.csv ~/data_sm/kmer-hashing/tabula-muris/n_hashes=500/bladder/similarities_tidy.csv

download: s3://kmer-hashing/tabula-muris/n_hashes=500/bladder/similarities_tidy.csv to ../../../../data_sm/kmer-hashing/tabula-muris/n_hashes=500/bladder/similarities_tidy.csv


In [3]:
similarities_tidy = pd.read_csv('s3://kmer-hashing/tabula-muris/n_hashes=500/bladder/similarities_tidy.csv')
print(similarities_tidy.shape)
similarities_tidy.head(10)

(1612900, 6)


cell1                    cell2  \
0  A1-B000610-3_56_F-1-1    A1-B000610-3_56_F-1-1   
1  A1-B000610-3_56_F-1-1    A1-B002764-3_38_F-1-1   
2  A1-B000610-3_56_F-1-1    A1-B002771-3_39_F-1-1   
3  A1-B000610-3_56_F-1-1     A1-D041914-3_8_M-1-1   
4  A1-B000610-3_56_F-1-1     A1-D042253-3_9_M-1-1   
5  A1-B000610-3_56_F-1-1  A1-MAA000487-3_10_M-1-1   
6  A1-B000610-3_56_F-1-1   A10-B000610-3_56_F-1-1   
7  A1-B000610-3_56_F-1-1   A10-B002764-3_38_F-1-1   
8  A1-B000610-3_56_F-1-1   A10-B002771-3_39_F-1-1   
9  A1-B000610-3_56_F-1-1    A10-D041914-3_8_M-1-1   

   similarity_without_abundance  similarity_with_abundance  \
0                         1.000                   1.000000   
1                         0.018                   0.228460   
2                         0.072                   0.245190   
3                         0.098                   0.454728   
4                         0.090                   0.231834   
5                         0.102                   0.216492   
6                         0.086                   0.093504   
7                         0.058                   0.079344   
8                         0.088                   0.374694   
9                         0.096                   0.181217   

  cell_ontology_class_cell1      cell_ontology_class  
0              bladder cell             bladder cell  
1              bladder cell  bladder urothelial cell  
2              bladder cell             bladder cell  
3              bladder cell             bladder cell  
4              bladder cell             bladder cell  
5              bladder cell             bladder cell  
6              bladder cell  bladder urothelial cell  
7              bladder cell  bladder urothelial cell  
8              bladder cell  bladder urothelial cell  
9              bladder cell             bladder cell

In [4]:
folder = '/home/olga/pureScratch/olgabot-maca/facs/sourmash/'


In [87]:
import json

from IPython.lib.pretty import pretty

with open(f"{folder}/A1-B000610-3_56_F-1-1_S28.sig") as f:
    signature_data = json.load(f)

print(pretty(signature_data, max_seq_length=10))

[{'class': 'sourmash_signature',
  'email': '',
  'filename': '/arg/2/0',
  'hash_function': '0.murmur64',
  'license': 'CC0',
  'name': 'A1-B000610-3_56_F-1-1_S28',
  'signatures': [{'abundances': [5, 1, 14, 32, 8, 2, 1, 1, 46, 1, ...],
    'ksize': 21,
    'max_hash': 0,
    'md5sum': '280b9babf31a3b7f820f9695f948808c',
    'mins': [321797371588609,
     224101973024771,
     113000222478351,
     1333158846758931,
     1102392644880415,
     587674942934704,
     586113718319139,
     278315472003108,
     875509323118629,
     399207939020840,
     ...],
    'molecule': 'DNA',
    'num': 500,
    'seed': 42},
   {'abundances': [7, 11, 1, 2, 1, 1, 2, 8, 5, 1, ...],
    'ksize': 31,
    'max_hash': 0,
    'md5sum': '9d82f6c550e83cdd97a9190e2fa7d1f0',
    'mins': [1098995281889281,
     494274597793794,
     805332503111685,
     944591797540874,
     433029384222735,
     777578192238617,
     815900320305157,
     729433322275847,
     593759110850609,
     823805458110523,
     ...

In [81]:
from IPython.display import JSON

JSON(signature_data)

<IPython.core.display.JSON object>

In [10]:
# bladder cell
cell1 = 'A1-B000610-3_56_F-1-1'

ksize = 21
moltype = 'DNA'


sig1 = list(sig.load_signatures(f"{folder}/A1-B000610-3_56_F-1-1_S28.sig", ksize=ksize, 
                                select_moltype=moltype))[0]
sig1

SourmashSignature('A1-B000610-3_56_F-1-1_S28', 280b9bab)

## Highly similar (jaccard = 0.102)  bladder cell

* Without abundance: 0.102
* With abundance: 0.216492

In [6]:
! ls -lha $folder/A1-MAA000487-3_10_M-1-1*

-rw-rw-r-- 1 olga olga 27K Jun 12  2018 /home/olga/pureScratch/olgabot-maca/facs/sourmash//A1-MAA000487-3_10_M-1-1_S264.sig


In [7]:
# bladder cell
cell1a = 'A1-MAA000487-3_10_M-1-1'
ksize = 21
moltype = 'DNA'


sig1a = list(sig.load_signatures(f"{folder}/A1-MAA000487-3_10_M-1-1_S264.sig", ksize=ksize, 
                                select_moltype=moltype))[0]
sig1a

SourmashSignature('A1-MAA000487-3_10_M-1-1_S264', eab086a5)

## Highly dissimilar (jaccard = 0.018) bladder urothelial cell

* Without abundance: 0.018
* With abundance: 0.228460

In [8]:
# bladder urothelial cell
cell2 = 'A1-B002764-3_38_F-1-1'

sig2 = list(sig.load_signatures(f"{folder}/A1-B002764-3_38_F-1-1_S291.sig", ksize=ksize, 
                                select_moltype=moltype))[0]
sig2

SourmashSignature('A1-B002764-3_38_F-1-1_S291', ef185816)

### Another dissimilar bladder urothelial cell

* Without abundance: 0.088
* With abundance: 0.374694

In [ ]:
! ls -lha $folder/A10-B002771-3_39_F-1-1*

In [ ]:
cell2a = 'A10-B002771-3_39_F-1-1'

sig2a = list(sig.load_signatures(f"{folder}/A10-B002771-3_39_F-1-1_S249.sig", ksize=ksize, 
                                select_moltype=moltype))[0]
sig2a

In [30]:
sig1_hashes = pd.Series(sig1.minhash.get_mins(with_abundance=True), name=cell1 + "__bladder_cell")
sig1_hashes.head()

3282762037724      2
5959368950693      1
7215285795798     20
11329527245923     1
12290763577842     1
Name: A1-B000610-3_56_F-1-1__bladder_cell, dtype: int64

In [31]:
sig1a_hashes = pd.Series(sig1a.minhash.get_mins(with_abundance=True), name=cell1a + "__bladder_cell")
sig1a_hashes.head()

1780366799926    2
3013374569767    1
3208195867768    1
3801525827063    1
4282618130625    1
Name: A1-MAA000487-3_10_M-1-1__bladder_cell, dtype: int64

In [32]:
sig2_hashes = pd.Series(sig2.minhash.get_mins(with_abundance=True), name=cell2 + "__bladder_urothelial_cell")
sig2_hashes.head()

4319888920071     1
11897683056067    4
18288931836992    1
27514404317028    1
28801856705165    1
Name: A1-B002764-3_38_F-1-1__bladder_urothelial_cell, dtype: int64

In [ ]:
2+2

In [ ]:
sig2a_hashes = pd.Series(sig2a.minhash.get_mins(with_abundance=True), name=cell2a + "__bladder_urothelial_cell")
sig2a_hashes.head()

In [33]:
sig1_hash_diff = (sig1_hashes - sig1a_hashes).dropna()
sig1_hash_diff

7215285795798      18.0
12290763577842    -14.0
13275177070369     -7.0
17454240377307     16.0
29190104568494      0.0
                   ... 
661085140555155    -4.0
666906505510494    11.0
670796309822097     5.0
695250169032813   -12.0
698576324166461    -3.0
Length: 66, dtype: float64

In [34]:
hash_diff = (sig1_hashes - sig2_hashes).dropna()
print(hash_diff.shape)
hash_diff

(14,)


171794956858086        1.0
394031160505449        0.0
513049510280006        5.0
635563321059591       54.0
734478954482604        0.0
790330283259107        0.0
860437213521678       -9.0
936637619733676       -4.0
937624453106451    -4675.0
1012839673718406    -240.0
1026077759632952    -276.0
1067106972354716       0.0
1215285019941982       0.0
1225906413461687       0.0
dtype: float64

In [154]:
dir(sig1)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_display_name',
 '_save',
 'contained_by',
 'd',
 'jaccard',
 'md5sum',
 'minhash',
 'name',
 'similarity']

In [35]:
combined_hashes = pd.concat([sig1_hashes, sig1a_hashes, sig2_hashes], axis=1)
# combined_hashes.columns = 
combined_hashes = combined_hashes.dropna(thresh=2)
print(combined_hashes.shape)
combined_hashes

(83, 3)


A1-B000610-3_56_F-1-1__bladder_cell  \
7215285795798                                    20.0   
12290763577842                                    1.0   
13275177070369                                    3.0   
17454240377307                                   17.0   
29190104568494                                    3.0   
...                                               ...   
1012839673718406                                  5.0   
1026077759632952                                 39.0   
1067106972354716                                  1.0   
1215285019941982                                  2.0   
1225906413461687                                  1.0   

                  A1-MAA000487-3_10_M-1-1__bladder_cell  \
7215285795798                                       2.0   
12290763577842                                     15.0   
13275177070369                                     10.0   
17454240377307                                      1.0   
29190104568494                                      3.0   
...                                                 ...   
1012839673718406                                    NaN   
1026077759632952                                    NaN   
1067106972354716                                    NaN   
1215285019941982                                    NaN   
1225906413461687                                    NaN   

                  A1-B002764-3_38_F-1-1__bladder_urothelial_cell  
7215285795798                                                NaN  
12290763577842                                               NaN  
13275177070369                                               NaN  
17454240377307                                               NaN  
29190104568494                                               NaN  
...                                                          ...  
1012839673718406                                           245.0  
1026077759632952                                           315.0  
1067106972354716                                             1.0  
1215285019941982                                             2.0  
1225906413461687                                             1.0  

[83 rows x 3 columns]

In [36]:
combined_hashes.to_csv("bladder_combined_hashes_table.csv")

In [37]:
pd.Series(combined_hashes.index).to_csv('bladder_combined_hashes.txt', index=False, header=False)

In [57]:
hashes_of_interest = combined_hashes.loc[(combined_hashes.abs() > 50).any(axis=1)]
print(hashes_of_interest.shape)
hashes_of_interest

(12, 3)


A1-B000610-3_56_F-1-1__bladder_cell  \
140738397883624                                 124.0   
209410885084946                                   5.0   
261433242471262                                  39.0   
279081595294430                                  45.0   
395041855771997                                  56.0   
572229171998306                                  71.0   
581669390169753                                 135.0   
608570483811428                                  19.0   
635563321059591                                  55.0   
937624453106451                                 301.0   
1012839673718406                                  5.0   
1026077759632952                                 39.0   

                  A1-MAA000487-3_10_M-1-1__bladder_cell  \
140738397883624                                    85.0   
209410885084946                                    54.0   
261433242471262                                   160.0   
279081595294430                                    67.0   
395041855771997                                    66.0   
572229171998306                                    21.0   
581669390169753                                   253.0   
608570483811428                                   100.0   
635563321059591                                    21.0   
937624453106451                                     NaN   
1012839673718406                                    NaN   
1026077759632952                                    NaN   

                  A1-B002764-3_38_F-1-1__bladder_urothelial_cell  
140738397883624                                              NaN  
209410885084946                                              NaN  
261433242471262                                              NaN  
279081595294430                                              NaN  
395041855771997                                              NaN  
572229171998306                                              NaN  
581669390169753                                              NaN  
608570483811428                                              NaN  
635563321059591                                              1.0  
937624453106451                                           4976.0  
1012839673718406                                           245.0  
1026077759632952                                           315.0

In [63]:
hashes_of_interest.to_csv('bladder_hashes_different_abundances_table.csv')

In [39]:
pwd

'/home/olga/code/kmer-hashing/kh-analysis/notebooks'

In [40]:
txt = "bladder_hashes_different_abundances.txt"

pd.Series(hashes_of_interest.index).to_csv(txt, index=False, header=False)

In [41]:
! cat $txt

140738397883624
209410885084946
261433242471262
279081595294430
395041855771997
572229171998306
581669390169753
608570483811428
635563321059591
937624453106451
1012839673718406
1026077759632952


In [42]:
fastqs = pd.read_csv("s3://czb-maca/Plate_seq/3_month/fastqs_read1_read2.csv", index_col=0)
print(fastqs.shape)
fastqs.head()

(53760, 2)


read1  \
sample_id                                                                  
A1-B000126-3_39_F-1-1  s3://czb-maca/Plate_seq/3_month/170925_A00111_...   
A1-B000127-3_38_F-1-1  s3://czb-maca/Plate_seq/3_month/170914_A00111_...   
A1-B000167-3_56_F-1-1  s3://czb-maca/Plate_seq/3_month/170925_A00111_...   
A1-B000168-3_57_F-1-1  s3://czb-maca/Plate_seq/3_month/170925_A00111_...   
A1-B000412-3_56_F-1-1  s3://czb-maca/Plate_seq/3_month/170914_A00111_...   

                                                                   read2  
sample_id                                                                 
A1-B000126-3_39_F-1-1  s3://czb-maca/Plate_seq/3_month/170925_A00111_...  
A1-B000127-3_38_F-1-1  s3://czb-maca/Plate_seq/3_month/170914_A00111_...  
A1-B000167-3_56_F-1-1  s3://czb-maca/Plate_seq/3_month/170925_A00111_...  
A1-B000168-3_57_F-1-1  s3://czb-maca/Plate_seq/3_month/170925_A00111_...  
A1-B000412-3_56_F-1-1  s3://czb-maca/Plate_seq/3_month/170914_A00111_...

In [45]:
# these_cell_fastqs = fastqs.loc[[cell1, cell1a, cell2]]
these_cell_fastqs = fastqs.loc[[cell1a]]
these_cell_fastqs

read1  \
sample_id                                                                    
A1-MAA000487-3_10_M-1-1  s3://czb-maca/Plate_seq/3_month/170910_A00111_...   

                                                                     read2  
sample_id                                                                   
A1-MAA000487-3_10_M-1-1  s3://czb-maca/Plate_seq/3_month/170910_A00111_...

### Copy fastqs over

In [46]:
fastq_dir = '/home/olga/pureScratch/czb-maca/Plate_Seq/3_month/170914_A00111_0058_AH3FYKDMXX__170914_A00111_0057_BH3FY7DMXX/fastqs/'

for fastq in these_cell_fastqs.values.flatten():
    print(fastq)
    ! aws s3 cp $fastq $fastq_dir

! ls $fastq_dir

s3://czb-maca/Plate_seq/3_month/170910_A00111_0054_AH2HGWDMXX__170910_A00111_0053_BH2HGKDMXX/fastqs/A1-MAA000487-3_10_M-1-1_R1_001.fastq.gz
download: s3://czb-maca/Plate_seq/3_month/170910_A00111_0054_AH2HGWDMXX__170910_A00111_0053_BH2HGKDMXX/fastqs/A1-MAA000487-3_10_M-1-1_R1_001.fastq.gz to ../../../../pureScratch/czb-maca/Plate_Seq/3_month/170914_A00111_0058_AH3FYKDMXX__170914_A00111_0057_BH3FY7DMXX/fastqs/A1-MAA000487-3_10_M-1-1_R1_001.fastq.gz
s3://czb-maca/Plate_seq/3_month/170910_A00111_0054_AH2HGWDMXX__170910_A00111_0053_BH2HGKDMXX/fastqs/A1-MAA000487-3_10_M-1-1_R2_001.fastq.gz
download: s3://czb-maca/Plate_seq/3_month/170910_A00111_0054_AH2HGWDMXX__170910_A00111_0053_BH2HGKDMXX/fastqs/A1-MAA000487-3_10_M-1-1_R2_001.fastq.gz to ../../../../pureScratch/czb-maca/Plate_Seq/3_month/170914_A00111_0058_AH3FYKDMXX__170914_A00111_0057_BH3FY7DMXX/fastqs/A1-MAA000487-3_10_M-1-1_R2_001.fastq.gz
A1-B000610-3_56_F-1-1_R1_001.fastq.gz  A1-B002764-3_38_F-1-1_R2_001.fastq.gz
A1-B000610-3_56_F-1

## Extract reads containing the hashes of interest

In [47]:
fastqs = glob.glob(f"{fastq_dir}*")
len(fastqs)
fastqs

['/home/olga/pureScratch/czb-maca/Plate_Seq/3_month/170914_A00111_0058_AH3FYKDMXX__170914_A00111_0057_BH3FY7DMXX/fastqs/A1-B000610-3_56_F-1-1_R1_001.fastq.gz',
 '/home/olga/pureScratch/czb-maca/Plate_Seq/3_month/170914_A00111_0058_AH3FYKDMXX__170914_A00111_0057_BH3FY7DMXX/fastqs/A1-B000610-3_56_F-1-1_R2_001.fastq.gz',
 '/home/olga/pureScratch/czb-maca/Plate_Seq/3_month/170914_A00111_0058_AH3FYKDMXX__170914_A00111_0057_BH3FY7DMXX/fastqs/A1-B002764-3_38_F-1-1_R1_001.fastq.gz',
 '/home/olga/pureScratch/czb-maca/Plate_Seq/3_month/170914_A00111_0058_AH3FYKDMXX__170914_A00111_0057_BH3FY7DMXX/fastqs/A1-B002764-3_38_F-1-1_R2_001.fastq.gz',
 '/home/olga/pureScratch/czb-maca/Plate_Seq/3_month/170914_A00111_0058_AH3FYKDMXX__170914_A00111_0057_BH3FY7DMXX/fastqs/A1-MAA000487-3_10_M-1-1_R1_001.fastq.gz',
 '/home/olga/pureScratch/czb-maca/Plate_Seq/3_month/170914_A00111_0058_AH3FYKDMXX__170914_A00111_0057_BH3FY7DMXX/fastqs/A1-MAA000487-3_10_M-1-1_R2_001.fastq.gz']

## Check if each sequence contains hashes, then extract them

In [ ]:
from sourmash._minhash import 

In [153]:
%%time
import sys
import string

query_hashes = set(hashes_of_interest.index)
minhash = sig1.minhash.copy_and_clear()


# now, iterate over the input sequences and output those that add
# hashes!
n = 0
m = 0

NOTIFY_EVERY_BP=1e6

reverse_complement_translation = str.maketrans("ACTG", "TGAC")

def get_reverse_complement(seq):
    return seq.translate(reverse_complement_translation)[::-1]

def canonical_kmer(kmer):
    rev_kmer = get_reverse_complement(kmer)
    if rev_kmer < kmer:
        return rev_kmer
    return kmer

all_hashes_of_interest_to_kmer = {}

watermark = NOTIFY_EVERY_BP

fasta = 'bladder_test_cells_different_hashes.fasta'

with open(fasta, 'w') as f:

    for fastq in fastqs:
        sample_id = os.path.basename(fastq).split("_R")[0]
        for record in screed.open(fastq):
            minhash = sig1.minhash.copy_and_clear()
            L = len(minhash.get_hashes())
            if L > 0:
                print(L)

            n += len(record.sequence)
            while n >= watermark:
                sys.stderr.write('... {} {}\r'.format(watermark, fastq))
                watermark += NOTIFY_EVERY_BP

            minhash.add_sequence(record.sequence, force=True)
            hashes = set(minhash.get_hashes())
            match = hashes.intersection(query_hashes)
            if match:
#                 break
                n_kmers = len(record.sequence) - ksize + 1
                kmers = [record.sequence[i:(i+ksize)] for i in range(n_kmers)]
                canonical_kmers = [canonical_kmer(kmer) for kmer in kmers]
                hashed_kmers = [hash_murmur(kmer) for kmer in canonical_kmers]
                hash_to_kmer = dict(zip(hashed_kmers, canonical_kmers))

                hashes_of_interest_to_kmer = {h: kmer for h, kmer in hash_to_kmer.items() if h in query_hashes}
                if len(hashes_of_interest_to_kmer) != len(match):
                    print("Something is weird! Not all the hashes were found...")
                    print(f"{match} were in the minhashes but not when individually hashing the k-mers")
                    break
                all_hashes_of_interest_to_kmer.update(hashes_of_interest_to_kmer)
            if len(all_hashes_of_interest_to_kmer) == len(query_hashes):
                print("Found all hashes! Exiting.")
                break
                
    for h, kmer in all_hashes_of_interest_to_kmer.items():
        f.write(f">{h}_canonical\n{kmer}\n")
        revcomp = get_reverse_complement(kmer)
        f.write(f">{h}_revcompl\n{kmer}\n")

! head $fasta
! wc -l $fasta

Found all hashes! Exiting.
Found all hashes! Exiting.
Found all hashes! Exiting.
Found all hashes! Exiting.
Found all hashes! Exiting.
Found all hashes! Exiting.
>581669390169753_canonical
CTGTAGTTGTGAACTGAGTTA
>581669390169753_revcompl
CTGTAGTTGTGAACTGAGTTA
>261433242471262_canonical
CACTTACTGTGTTCCTATGGA
>261433242471262_revcompl
CACTTACTGTGTTCCTATGGA
>395041855771997_canonical
ACTTTATCACCTGGTTTCACA
48 bladder_test_cells_different_hashes.fasta
CPU times: user 16.7 s, sys: 28 ms, total: 16.7 s
Wall time: 17.1 s


In [150]:
! wc -l bladder_test_cells_different_hashes.fasta

24 bladder_test_cells_different_hashes.fasta


In [151]:
cat bladder_test_cells_different_hashes.fasta

>581669390169753
CTGTAGTTGTGAACTGAGTTA
>261433242471262
CACTTACTGTGTTCCTATGGA
>395041855771997
ACTTTATCACCTGGTTTCACA
>937624453106451
ATCGAGGTCATGACTGGACAC
>140738397883624
GACCCCGCCGATGTCGCTATC
>279081595294430
CATCTTGCTAACAAGTTTGAA
>635563321059591
AGTGGACATGAAAGATGTGGA
>1012839673718406
GGAGTGGGTAAATTGCGCGCC
>572229171998306
AAGCCTATCTGGACCAGAAAC
>1026077759632952
GCTCAGTTATTGACTACGAGC
>608570483811428
AGGAAGCTGGGGTGTTGTGGG
>209410885084946
AAAGAAATAGACACTAATGGA


In [139]:
all_hashes_of_interest_to_kmer

{581669390169753: 'CTGTAGTTGTGAACTGAGTTA',
 261433242471262: 'CACTTACTGTGTTCCTATGGA',
 395041855771997: 'ACTTTATCACCTGGTTTCACA',
 937624453106451: 'ATCGAGGTCATGACTGGACAC',
 140738397883624: 'GACCCCGCCGATGTCGCTATC',
 279081595294430: 'CATCTTGCTAACAAGTTTGAA',
 635563321059591: 'AGTGGACATGAAAGATGTGGA',
 1012839673718406: 'GGAGTGGGTAAATTGCGCGCC',
 572229171998306: 'AAGCCTATCTGGACCAGAAAC',
 1026077759632952: 'GCTCAGTTATTGACTACGAGC',
 608570483811428: 'AGGAAGCTGGGGTGTTGTGGG',
 209410885084946: 'AAAGAAATAGACACTAATGGA'}

In [136]:
all_hashes_of_interest_to_kmer

{581669390169753: 'CTGTAGTTGTGAACTGAGTTA',
 261433242471262: 'TCCATAGGAACACAGTAAGTG',
 395041855771997: 'TGTGAAACCAGGTGATAAAGT',
 937624453106451: 'GTGTCCAGTCATGACCTCGAT',
 140738397883624: 'GATAGCGACATCGGCGGGGTC',
 279081595294430: 'TTCAAACTTGTTAGCAAGATG',
 635563321059591: 'AGTGGACATGAAAGATGTGGA',
 1012839673718406: 'GGAGTGGGTAAATTGCGCGCC',
 572229171998306: 'AAGCCTATCTGGACCAGAAAC',
 1026077759632952: 'GCTCGTAGTCAATAACTGAGC',
 608570483811428: 'CCCACAACACCCCAGCTTCCT',
 209410885084946: 'AAAGAAATAGACACTAATGGA'}

In [ ]:
hashes_of_interest.index.name = 'hash_murmur'

In [152]:
hashes_of_interest['kmer'] = pd.Series(all_hashes_of_interest_to_kmer)
print(hashes_of_interest.shape)
hashes_of_interest

(12, 4)


/home/olga/miniconda3/envs/kmer-hashing/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


A1-B000610-3_56_F-1-1__bladder_cell  \
hash_murmur                                             
140738397883624                                 124.0   
209410885084946                                   5.0   
261433242471262                                  39.0   
279081595294430                                  45.0   
395041855771997                                  56.0   
572229171998306                                  71.0   
581669390169753                                 135.0   
608570483811428                                  19.0   
635563321059591                                  55.0   
937624453106451                                 301.0   
1012839673718406                                  5.0   
1026077759632952                                 39.0   

                  A1-MAA000487-3_10_M-1-1__bladder_cell  \
hash_murmur                                               
140738397883624                                    85.0   
209410885084946                                    54.0   
261433242471262                                   160.0   
279081595294430                                    67.0   
395041855771997                                    66.0   
572229171998306                                    21.0   
581669390169753                                   253.0   
608570483811428                                   100.0   
635563321059591                                    21.0   
937624453106451                                     NaN   
1012839673718406                                    NaN   
1026077759632952                                    NaN   

                  A1-B002764-3_38_F-1-1__bladder_urothelial_cell  \
hash_murmur                                                        
140738397883624                                              NaN   
209410885084946                                              NaN   
261433242471262                                              NaN   
279081595294430                                              NaN   
395041855771997                                              NaN   
572229171998306                                              NaN   
581669390169753                                              NaN   
608570483811428                                              NaN   
635563321059591                                              1.0   
937624453106451                                           4976.0   
1012839673718406                                           245.0   
1026077759632952                                           315.0   

                                   kmer  
hash_murmur                              
140738397883624   GACCCCGCCGATGTCGCTATC  
209410885084946   AAAGAAATAGACACTAATGGA  
261433242471262   CACTTACTGTGTTCCTATGGA  
279081595294430   CATCTTGCTAACAAGTTTGAA  
395041855771997   ACTTTATCACCTGGTTTCACA  
572229171998306   AAGCCTATCTGGACCAGAAAC  
581669390169753   CTGTAGTTGTGAACTGAGTTA  
608570483811428   AGGAAGCTGGGGTGTTGTGGG  
635563321059591   AGTGGACATGAAAGATGTGGA  
937624453106451   ATCGAGGTCATGACTGGACAC  
1012839673718406  GGAGTGGGTAAATTGCGCGCC  
1026077759632952  GCTCAGTTATTGACTACGAGC

In [148]:
print(hashes_of_interest.to_csv(sep='\t'))


hash_murmur	A1-B000610-3_56_F-1-1__bladder_cell	A1-MAA000487-3_10_M-1-1__bladder_cell	A1-B002764-3_38_F-1-1__bladder_urothelial_cell	kmer
140738397883624	124.0	85.0		GACCCCGCCGATGTCGCTATC
209410885084946	5.0	54.0		AAAGAAATAGACACTAATGGA
261433242471262	39.0	160.0		CACTTACTGTGTTCCTATGGA
279081595294430	45.0	67.0		CATCTTGCTAACAAGTTTGAA
395041855771997	56.0	66.0		ACTTTATCACCTGGTTTCACA
572229171998306	71.0	21.0		AAGCCTATCTGGACCAGAAAC
581669390169753	135.0	253.0		CTGTAGTTGTGAACTGAGTTA
608570483811428	19.0	100.0		AGGAAGCTGGGGTGTTGTGGG
635563321059591	55.0	21.0	1.0	AGTGGACATGAAAGATGTGGA
937624453106451	301.0		4976.0	ATCGAGGTCATGACTGGACAC
1012839673718406	5.0		245.0	GGAGTGGGTAAATTGCGCGCC
1026077759632952	39.0		315.0	GCTCAGTTATTGACTACGAGC



In [ ]:
hashes_of_interest.to_csv

In [67]:
# kmers

In [68]:
manual_hash_with_minhash = set(hashed_kmers) & set(minhash.get_hashes())
len(manual_hash_with_minhash)

0

In [69]:
intersecting_hashes = hash_to_kmer_series[list(manual_hash_with_minhash)]
print(intersecting_hashes.shape)
intersecting_hashes.head()

(0,)


Series([], dtype: object)

In [70]:
intersecting_hashes.sort_values()

Series([], dtype: object)

In [71]:
len('TGGACACTGCATCGGAAGACA')

21

In [47]:
record.sequence

'AGCAGGGGGCGTCCCTATACGCGATGGATCGAGGTCATGACTGGACACTGCATCGGAAGACACTTCGTGCCGACACGAAGCAGCCTCAATATCTAACACG'

In [53]:
intersecting_hashes_df = intersecting_hashes.to_frame()
intersecting_hashes_df = intersecting_hashes_df.rename(columns={0: 'kmer'})
intersecting_hashes_df.head()

kmer
2.267453e+18  CCCTATACGCGATGGATCGAG
4.345891e+18  CCTATACGCGATGGATCGAGG
1.434296e+19  AGACACTTCGTGCCGACACGA
1.013144e+19  CATGACTGGACACTGCATCGG
7.240080e+18  GGCGTCCCTATACGCGATGGA

In [54]:
intersecting_hashes_df['start_index'] = intersecting_hashes_df.kmer.map(lambda x: record.sequence.find(x))
intersecting_hashes_df.head()

kmer  start_index
2.267453e+18  CCCTATACGCGATGGATCGAG           12
4.345891e+18  CCTATACGCGATGGATCGAGG           13
1.434296e+19  AGACACTTCGTGCCGACACGA           57
1.013144e+19  CATGACTGGACACTGCATCGG           35
7.240080e+18  GGCGTCCCTATACGCGATGGA            7

In [56]:
len(record.sequence)

100

In [55]:
intersecting_hashes_df.sort_values('start_index')

kmer  start_index
9.882987e+18  AGCAGGGGGCGTCCCTATACG            0
1.674263e+19  GCAGGGGGCGTCCCTATACGC            1
9.725331e+18  CAGGGGGCGTCCCTATACGCG            2
1.204699e+19  AGGGGGCGTCCCTATACGCGA            3
7.240080e+18  GGCGTCCCTATACGCGATGGA            7
1.766957e+19  CGTCCCTATACGCGATGGATC            9
2.411537e+18  TCCCTATACGCGATGGATCGA           11
2.267453e+18  CCCTATACGCGATGGATCGAG           12
4.345891e+18  CCTATACGCGATGGATCGAGG           13
4.573475e+18  ATACGCGATGGATCGAGGTCA           16
7.878499e+18  ACGCGATGGATCGAGGTCATG           18
1.269715e+19  CGCGATGGATCGAGGTCATGA           19
5.852042e+17  GCGATGGATCGAGGTCATGAC           20
1.081915e+19  ATGGATCGAGGTCATGACTGG           23
2.460388e+18  GGATCGAGGTCATGACTGGAC           25
1.569262e+19  GATCGAGGTCATGACTGGACA           26
9.376245e+14  ATCGAGGTCATGACTGGACAC           27
1.733795e+19  GAGGTCATGACTGGACACTGC           30
7.272113e+18  AGGTCATGACTGGACACTGCA           31
1.013144e+19  CATGACTGGACACTGCATCGG           35
1.486561e+19  ATGACTGGACACTGCATCGGA           36
1.553129e+19  TGACTGGACACTGCATCGGAA           37
1.201782e+18  ACTGGACACTGCATCGGAAGA           39
3.912292e+18  CTGGACACTGCATCGGAAGAC           40
1.401186e+19  TGGACACTGCATCGGAAGACA           41
6.281666e+18  GGACACTGCATCGGAAGACAC           42
1.438740e+19  CACTGCATCGGAAGACACTTC           45
1.458269e+19  ACTGCATCGGAAGACACTTCG           46
2.033339e+18  CATCGGAAGACACTTCGTGCC           50
1.102780e+19  ATCGGAAGACACTTCGTGCCG           51
1.653177e+19  TCGGAAGACACTTCGTGCCGA           52
8.229911e+18  CGGAAGACACTTCGTGCCGAC           53
3.914028e+18  GGAAGACACTTCGTGCCGACA           54
1.606500e+18  GAAGACACTTCGTGCCGACAC           55
3.900669e+17  AAGACACTTCGTGCCGACACG           56
1.434296e+19  AGACACTTCGTGCCGACACGA           57
7.956584e+18  GACACTTCGTGCCGACACGAA           58
1.197956e+19  ACACTTCGTGCCGACACGAAG           59
8.438121e+18  CACTTCGTGCCGACACGAAGC           60
6.856428e+18  ACTTCGTGCCGACACGAAGCA           61
2.301512e+18  GCCGACACGAAGCAGCCTCAA           68
1.652119e+19  CGACACGAAGCAGCCTCAATA           70
4.062782e+18  ACACGAAGCAGCCTCAATATC           72
1.652205e+19  ACGAAGCAGCCTCAATATCTA           74
6.682628e+18  CGAAGCAGCCTCAATATCTAA           75
1.473452e+19  GAAGCAGCCTCAATATCTAAC           76
5.343365e+18  AAGCAGCCTCAATATCTAACA           77
6.531533e+18  AGCAGCCTCAATATCTAACAC           78

In [49]:
for kmer in intersecting_hashes.values:
    print(record.sequence.find(kmer))

12
13
57
35
7
27
26
11
30
46
50
18
78
25
0
76
23
60
42
36
53
74
40
77
68
31
58
41
51
61
55
16
75
19
45
2
20
72
70
9
56
37
39
1
52
59
3
54


In [36]:
# manual_hash_with_minhash

In [32]:
len(set(minhash.get_hashes()))

80

In [34]:
minhash.add_sequence(record.sequence, force=True)
len(set(minhash.get_hashes()))

80

In [31]:
len(set(hashed_kmers))

80

In [25]:
hashes_of_interest_to_kmer

{}

In [26]:
kmer_to_hash

{'GTATCAACAGGAACATCCTCG': 11332569200613985422,
 'TATCAACAGGAACATCCTCGT': 13631614553446097611,
 'ATCAACAGGAACATCCTCGTG': 2872127591073786859,
 'TCAACAGGAACATCCTCGTGT': 7665439403394214361,
 'CAACAGGAACATCCTCGTGTT': 10755392401481311365,
 'AACAGGAACATCCTCGTGTTA': 16116729706801105248,
 'ACAGGAACATCCTCGTGTTAG': 4928222811555832491,
 'CAGGAACATCCTCGTGTTAGA': 497161709746860352,
 'AGGAACATCCTCGTGTTAGAT': 17807558691675200061,
 'GGAACATCCTCGTGTTAGATA': 10713124010397424574,
 'GAACATCCTCGTGTTAGATAT': 3812529855415123365,
 'AACATCCTCGTGTTAGATATT': 15055893384930568398,
 'ACATCCTCGTGTTAGATATTG': 15090075336066576766,
 'CATCCTCGTGTTAGATATTGA': 9761570019713904088,
 'ATCCTCGTGTTAGATATTGAG': 4424664125813896497,
 'TCCTCGTGTTAGATATTGAGG': 3215167914112085868,
 'CCTCGTGTTAGATATTGAGGC': 16533125832238829217,
 'CTCGTGTTAGATATTGAGGCT': 5199017375569319483,
 'TCGTGTTAGATATTGAGGCTG': 11855670425143298460,
 'CGTGTTAGATATTGAGGCTGC': 5401960134113432925,
 'GTGTTAGATATTGAGGCTGCT': 16924423555563794259,
 'T